src/agent.rs

```diff
@@ -0,0 +1,574 @@
+impl RxAgent {
+    /// Start an `RxAgent`, spawn a thread to do the polling job.
+    pub(crate) fn start(socket_id: i32) -> Arc<Self> {
+        let running = AtomicBool::new(true);
+        let this = Arc::new(RxAgent {
+            running,
+            tasks: Mutex::new(BTreeSet::new()),
+        });
+        let that = Arc::clone(&this);
+        let _handle = task::spawn_blocking(move || {
+            let mut frag_tbl = IpFragmentTable::new(socket_id)?;
+            let mut death_row = IpFragDeathRow::new(socket_id)?;
+
+            while that.running.load(Ordering::Acquire) {
+                let tasks = that.tasks.lock().map_err(Error::from)?;
+                let task_iter = tasks.iter();
+                for &(port_id, queue_id) in task_iter {
+                    let mut ptrs = vec![ptr::null_mut(); MAX_PKT_BURST as usize];
+                    // SAFETY: ffi
+                    let n = unsafe {
+                        rte_eth_rx_burst(port_id, queue_id, ptrs.as_mut_ptr(), MAX_PKT_BURST)
+                    };
+                    trace!("{n} packets received");
+                    for ptr in ptrs.into_iter().take(n as _) {
+                        let m = Mbuf::new_with_ptr(ptr)?;
+                        if let Some((sockfd, res)) = handle_ether(m, &mut frag_tbl, &mut death_row)
+                        {
+                            let res = socket::put_mailbox(sockfd, res);
+                            if let Err(e) = res {
+                                error!("An error {e} occurred in `put_mailbox`");
+                            }
+                        }
+                    }
+                }
+            }
+            info!("RxAgent thread terminated");
+            Result::<()>::Ok(())
+        });
+        this
+    }
+
+    /// Stop the `RxAgent`.
+    pub(crate) fn stop(self: &Arc<Self>) {
+        self.running.store(false, Ordering::Release);
+    }
+
+    /// Register a (`port_id`, `queue_id`) to an `RxAgent`.
+    ///
+    /// Adds a (`port_id`, `queue_id`) pair to the set to be polled.
+    pub(crate) fn register(self: &Arc<Self>, port_id: u16, queue_id: u16) -> Result<()> {
+        let _ = self
+            .tasks
+            .lock()
+            .map_err(Error::from)?
+            .insert((port_id, queue_id));
+        Ok(())
+    }
+
+    /// Unregister a (`port_id`, `queue_id`) from an `RxAgent`.
+    ///
+    /// Removes the (`port_id`, `queue_id`) pair from the polled set.
+    pub(crate) fn unregister(self: &Arc<Self>, port_id: u16, queue_id: u16) -> Result<()> {
+        let _ = self
+            .tasks
+            .lock()
+            .map_err(Error::from)?
+            .remove(&(port_id, queue_id));
+        Ok(())
+    }
+}
+
+#[allow(unsafe_code)]
+impl TxAgent {
+    /// Start a `TxBuffer`, spawn a thread to do the sending job.
+    pub(crate) fn start() -> Arc<Self> {
+        #[allow(clippy::unwrap_used)]
+        let rt = Builder::new_multi_thread()
+            .worker_threads(1)
```

Comment:
If the thread number is always 1, why are we using multi thread pool?

Reply:
At first, I used a `new_current_thread` and it just stuck here. I guess there're some different mechanisms behind `new_current_thread` and `new_multi_thread`?

Reply:
I suggest to investigate the mechanism behind the thread pool instead of using multi_thread pool. Single thread pool has some optimization that the other doesn't have.

Reply:
The multi-threaded Runtime uses a `Launch` to [spawn the thread workers](https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/builder.rs#L1060), which will [spawn all the worker threads](https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/scheduler/multi_thread/worker.rs#L363) due to configuration. However, a single thread Runtime does not use a `Launch` and just runs the Tasks on the current thread.

I will try to spawn a new thread and use a single thread Runtime on it.